File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/RegularMatchDeserializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;
import it.unibo.coordination.linda.text.RegexTemplate;
import it.unibo.coordination.linda.text.RegularMatch;
import it.unibo.coordination.linda.text.StringTuple;

import java.util.Map;

class RegularMatchDeserializer extends DynamicDeserializer<RegularMatch> {

    public RegularMatchDeserializer(MIMETypes mimeType, ObjectMapper mapper) {
        super(RegularMatch.class, mimeType, mapper);
    }

    @Override
    public RegularMatch fromDynamicObject(Object dynamicObject) {
        if (dynamicObject instanceof Map) {
            final Map<String, ?> dynamicMap = (Map<String, ?>) dynamicObject;

            if (dynamicMap.get("template") instanceof String) {
                final RegexTemplate template = RegexTemplate.of((String) dynamicMap.get("template"));

                final Object tupleObject;
                if ((tupleObject = dynamicMap.get("tuple")) instanceof String) {
                    final StringTuple tuple = StringTuple.of((String) tupleObject);

                    return template.matchWith(tuple);
                }

                return RegularMatch.failed(template);
            }
        }
        throw new IllegalArgumentException("Cannot read " + getSupportedMIMEType());
    }
}

File: linda-text-presentation/src/main/kotlin/it/unibo/coordination/linda/text/RegularMatchDeserializer.kt
package it.unibo.coordination.linda.text

import com.fasterxml.jackson.databind.ObjectMapper
import it.unibo.presentation.DynamicDeserializer
import it.unibo.presentation.MIMETypes

internal class RegularMatchDeserializer(mimeType: MIMETypes, mapper: ObjectMapper) : DynamicDeserializer<RegularMatch>(RegularMatch::class.java, mimeType, mapper) {
    override fun fromDynamicObject(dynamicObject: Any): RegularMatch {
        if (dynamicObject is Map<*, *>) {
            when (val templateStr = dynamicObject["template"] ) {
                is String -> {
                    val template = RegexTemplate.of(templateStr)
                    when(val tupleObject: Any? = dynamicObject["tuple"]) {
                        is String -> {
                            val tuple = StringTuple.of(tupleObject)
                            return template.matchWith(tuple)
                        }
                    }
                    return RegularMatch.failed(template)
                }
            }
        }
        throw IllegalArgumentException("Cannot read $supportedMIMEType")
    }
}